plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/concurrent/LSPExecutorService.java

```diff
@@ -0,0 +1,296 @@
+	public synchronized <T> QueuedTaskFuture<T> submit(Object queueId, String description,
+			Function<CancelIndicator, T> task) {
+		QueuedTask<T> queuedTask = createQueuedTask(queueId, description, task);
+		enqueue(queuedTask);
+		doSubmitAllPending();
+		return queuedTask.result;
+	}
+
+	/** Put the given task on the queue of pending tasks. */
+	protected synchronized void enqueue(QueuedTask<?> task) {
+		pendingTasks.add(task);
+	}
+
+	/** Remove and return the next pending, non-blocked task from the queue of pending tasks. */
+	protected synchronized QueuedTask<?> pollNextPending() {
+		Iterator<QueuedTask<?>> iter = pendingTasks.iterator();
+		while (iter.hasNext()) {
+			QueuedTask<?> curr = iter.next();
+			boolean isBlocked = runningTasks.containsKey(curr.queueId);
+			if (!isBlocked) {
+				iter.remove();
+				return curr;
+			}
+		}
+		return null;
+	}
+
+	/** Like {@link #doSubmit(QueuedTask)}, but will submit all currently pending, non-blocked tasks. */
+	protected synchronized void doSubmitAllPending() {
+		QueuedTask<?> next;
+		while ((next = pollNextPending()) != null) {
+			doSubmit(next);
+		}
+	}
+
+	/** Submit the given task to the delegate executor service. */
+	protected synchronized void doSubmit(QueuedTask<?> task) {
+		if (runningTasks.putIfAbsent(task.queueId, task) != null) {
+			throw new IllegalStateException("executor inconsistency: queue ID already in progress: " + task.queueId);
+		}
+		delegate.submit(task); // will eventually invoke #onDone()
+	}
+
+	/** Invoked by each running task upon completion, see {@link QueuedTask#run()}. */
+	protected synchronized void onDone(QueuedTask<?> task) {
+		if (runningTasks.remove(task.queueId) == null) {
+			throw new IllegalStateException("executor inconsistency: queue ID not in progress: " + task.queueId);
+		}
+		doSubmitAllPending();
+	}
+
+	/**
+	 * Marks all running and pending tasks as cancelled, i.e. their cancel indicator will return <code>true</code> from
+	 * {@link CancelIndicator#isCanceled() #isCanceled()}. Does not mark the tasks' result futures as cancelled, see
+	 * {@link LSPExecutorService} for details.
+	 */
+	public synchronized void cancelAll() {
+		Stream.concat(runningTasks.values().stream(), pendingTasks.stream())
+				.forEach(t -> t.cancel());
+	}
+
+	/** Same as {@link #cancelAll()}, but only affects tasks with a queue ID equal to the given ID. */
+	public synchronized void cancelAll(Object queueId) {
+		Stream.concat(runningTasks.values().stream(), pendingTasks.stream())
+				.filter(t -> queueId.equals(t.queueId))
+				.forEach(t -> t.cancel());
+	}
+
+	/**
+	 * Blocks until all tasks complete that are running or pending at the time of invocation of this method OR are being
+	 * submitted by other threads while this method is waiting. Thus, this method waits for this executor to idle.
+	 */
+	public /* NOT synchronized */ void join() {
+		CompletableFuture<Void> allTasks;
+		while ((allTasks = allTasksOrNull()) != null) {
```

Comment:
I don't think the JavaDoc is correct.
`allTasksOrNull may return null` and immediately thereafter another thread will submit a task. In that sense, the executor is not idle when `join` returns.